Fix: 화장실 검색 (#7), Feat: 리뷰 작성 (#9), Feat: 포인트 충전 (#10)

hyuntae99 opened the pull request:
### 🔖작업 내용

1. 리뷰 작성 구현 (하는 김에 필요할 것 같은 것들도 구현했음)
a. 리뷰 작성
b. 특정 리뷰 수정
c. 특정 리뷰 삭제
d. 특정 리뷰 조회
e. 해당 화장실에 대한 모든 리뷰 조회 (최신순, 별점 높은순, 별점 낮은순)
f. 화장실에 대한 평균 평점 조회 

2. 오류, 응답 통일 (ApiStandardResponse)
3. ExceptionHandler 분리
4. Swagger 코드 수정


### 🔖논의 사항 및 문제 
1. restroom에 평균 평점을 넣어야 하나 
(위의 api를 내부 함수로 바꾸고 리뷰 변동이 있을 때마다 해당 화장실 평균 평점을 업데이트하는 식으로 하면 될 듯)

2. GlobalHandler에 어떤 예외들이 들어가야 하나 (SwaggerConfig에 있는 예외만 넣었음 일단)

3. SecurityConfig에서  마지막 antMatcher의 엔드 포인트만 적용이 되는 문제 발견함 
(antMatchers 라는 방법도 있길래 써보니까 오류가 발생해서 일단 알아보는 중)

4. API 테스트할 때, 아까 수정한 User, Restroom에서 알 수 없는 문제가 발생해서 일단 제외했음
(일단은 당장 내 이슈 구현할 때는 문제가 없을 것 같아서 제외했어. 나중에 해결해볼게)

### 📨관련 이슈

Resolved: #7, #9, #10

hyuntae99 committed:
주석 수정

hyuntae99 committed:
오류 해결을 위한 설정 변경

hyuntae99 committed:
@Parameter 수정

hyuntae99 committed:
@Parameter 수정

hyuntae99 committed:
1. 리뷰 작성을 위한 controller, service, dto, repository 추가
2. 리뷰 작성에 대한 error code, exception class 추가
3. swagger 설정에 대한 코드 추가

hyuntae99 committed:
1. ErrorCode 상세화
2. Response 통일

hyuntae99 committed:
ApiStandardResponse로 응답을 통일

hyuntae99 committed:
ValidationException 추가

hyuntae99 committed:
1. ErrorCode 상세화
2. Response 통일

hyuntae99 committed:
GlobalExceptionHandler 수정

hyuntae99 committed:
ReviewController ApiStandardResponse 형식으로 응답

hyuntae99 committed:
ErrorCode 형식 수정

hyuntae99 committed:
merge local develop

hyuntae99 committed:
1. ExceptionHandler 세분화
2. Review 관련 코드에 대한 추가 및 수정

hyuntae99 committed:
upstream merge

hyuntae99 committed:
주석 추가

hyuntae99 committed:
.antMatcher 는 뒤에 있는 엔드포인트로 덮어씌워지므로 유의

hyuntae99 committed:
엔드포인트 수정

hyuntae99 committed:
엔드포인트 수정

hyuntae99 committed:
antMatcher 추가

hyuntae99 committed:
함수명 수정

hyuntae99 committed:
antMatcher 수정

hyuntae99 committed:
포인트 충전 구현

seokho-1116 commented:
> ### 🔖작업 내용
> 1. 리뷰 작성 구현 (하는 김에 필요할 것 같은 것들도 대충 구현해봤음...)
>    a. 리뷰 작성
>    b. 특정 리뷰 수정
>    c. 특정 리뷰 삭제
>    d. 특정 리뷰 조회
>    e. 해당 화장실에 대한 모든 리뷰 조회 (최신순, 별점순)
>    f. 화장실에 대한 평균 평점 조회
> 2. 오류, 응답 통일 (ApiStandardResponse)
> 3. ExceptionHandler 분리
> 4. Swagger 코드 수정
> 
> ### 🔖논의 사항 및 문제
> 1. restroom에 평균 평점을 넣어야 하나
>    (위의 api를 내부 함수로 바꾸고 리뷰 변동이 있을 때마다 해당 화장실 평균 평점을 업데이트하는 식으로 하면 될 듯)
> 2. GlobalHandler에 어떤 예외들이 들어가야 하나 (SwaggerConfig에 있는 예외만 넣었음 일단)
> 3. SecurityConfig에서  마지막 antMatcher의 엔드 포인트만 적용이 되는 문제 발견함
>    (antMatchers 라는 방법도 있길래 써보니까 오류가 발생해서 일단 알아보는 중)
> 4. API 테스트할 때, 아까 수정한 User, Restroom에서 알 수 없는 문제가 발생해서 일단 제외했음
>    (일단은 당장 내 이슈 구현할 때는 문제가 없을 것 같아서 제외했어. 나중에 해결해볼게)
> 
> ### 📨관련 이슈
> Resolved: #7, #9

3. 해결중이야 좀만 기둘기둘

seokho-1116 commented on src/main/java/com/project/chamjimayo/service/UserService.java:
```diff
@@ -44,4 +45,22 @@ public String getUserByAuthId(String authId) {
+    int currentPoint = user.getPoint() != null ? user.getPoint() : 0;
+    int newPoint = requestDTO.getPoint();
+    int totalPoint = newPoint + currentPoint;
+    user.changePoint(totalPoint);
```
changePoint 말고 addPoint해서 로직을 충전을 더하는 로직을 User안에 넣는게 더 좋아보임

hyuntae99 committed:
주석 수정

hyuntae99 committed:
upstream merge

hyuntae99 committed:
antMatchers 도입 (여러 엔드포인트 적용 완료)

hyuntae99 commented:
> 

해결했어!

seokho-1116 commented on src/main/java/com/project/chamjimayo/repository/UserRepository.java:
querydsl 도입으로 UserQueryRepository, UserJpaRepository로 분리했어 그래서 이름 바꿔줘야 할듯?

hyuntae99 replied:
일단 이름 UserRepository에서 UserJpaRepository 변경했고 나중에 merge하면서 추가 수정하면 될 듯

seokho-1116 commented:
> ### 🔖작업 내용
> 1. 리뷰 작성 구현 (하는 김에 필요할 것 같은 것들도 대충 구현해봤음...)
>    a. 리뷰 작성
>    b. 특정 리뷰 수정
>    c. 특정 리뷰 삭제
>    d. 특정 리뷰 조회
>    e. 해당 화장실에 대한 모든 리뷰 조회 (최신순, 별점순)
>    f. 화장실에 대한 평균 평점 조회
> 2. 오류, 응답 통일 (ApiStandardResponse)
> 3. ExceptionHandler 분리
> 4. Swagger 코드 수정
> 
> ### 🔖논의 사항 및 문제
> 1. restroom에 평균 평점을 넣어야 하나
>    (위의 api를 내부 함수로 바꾸고 리뷰 변동이 있을 때마다 해당 화장실 평균 평점을 업데이트하는 식으로 하면 될 듯)
> 2. GlobalHandler에 어떤 예외들이 들어가야 하나 (SwaggerConfig에 있는 예외만 넣었음 일단)
> 3. SecurityConfig에서  마지막 antMatcher의 엔드 포인트만 적용이 되는 문제 발견함
>    (antMatchers 라는 방법도 있길래 써보니까 오류가 발생해서 일단 알아보는 중)
> 4. API 테스트할 때, 아까 수정한 User, Restroom에서 알 수 없는 문제가 발생해서 일단 제외했음
>    (일단은 당장 내 이슈 구현할 때는 문제가 없을 것 같아서 제외했어. 나중에 해결해볼게)
> 
> ### 📨관련 이슈
> Resolved: #7, #9

1. 좋은듯
2. db오류, 서버오류, 스프링 controller 동작 과정 오류? 이런거 더 찾아보면 나올듯
3. 해결완료
4. 이거는 왜그런지 모르겠다?

hyuntae99 committed:
주석 정리

hyuntae99 committed:
1. Restroom에 averageRating 추가
2. Controller에 해당 api 삭제
3. 평점 계산 함수가 리뷰 추가,수정,삭제할 때마다 실행되도록 변경

hyuntae99 committed:
changePoint -> addPoint로 변경

hyuntae99 committed:
SecurityConfig 수정

hyuntae99 committed:
averageRating 기본 값 설정

hyuntae99 committed:
@Transactional 추가

hyuntae99 commented:
1. restroom에 average_rating을 추가하고 리뷰 추가, 수정, 삭제할 때마다 평균 평점을 계산하도록 구현
2. 일단 기본적인 예외만 넣었고 하면서 겹치는 오류들을 넣으면 될 것 같음.
3. 해결함 (테스트도 했어)
4. upstream merge 되면서 문제 해결된 듯?

hyuntae99 committed:
UserRepository -> UserJpaRepository 이름 변경

hyuntae99 committed:
1. 포인트 차감 기능 구현
2. Error 코드 추가

hyuntae99 committed:
PointChargeDto, PointDeductionDto 통합

hyuntae99 committed:
1. 포인트 충전, 차감 엔드포인트 변경
2. UserExceptionHandler에 추가된 예외 적용
3. Swagger에 변경사항 적용
4. SecurityConfig에 엔드포인트 추가

hyuntae99 committed:
포인트 충전/차감 코드 수정

hyuntae99 committed:
리뷰 리스트 조회 변경
(기존에 최신 리뷰가 앞으로 오도록 수정)

hyuntae99 commented:
1. 포인트 충전, 차감 구현 (요청 값이 같을 것 같아서 같은 dto로 처리)
2. ErrorCode 추가 (노션에 반영했어)
3. 기존 리뷰 리스트를 반환하면 최신 리뷰가 뒤로 가던 것을 최신순으로 볼 수 있도록 수정

seokho-1116 commented on src/main/java/com/project/chamjimayo/controller/ReviewController.java:
```diff
@@ -0,0 +1,260 @@
+	@Parameters({
+		@Parameter(in = ParameterIn.HEADER, name = "Bearer-Token", required = true)
+	})
+	@PostMapping("/write")
```
restful api에 대한 경로를 다는 법을 보면 좋을 듯?
동작이 아니라 자원 그 자체를 보고 엔드포인트를 만듬
/review
http메소드가 그 동작을 설명해준다.
[rest api](https://yangbongsoo.gitbook.io/study/undefined-1/rest)
[rest api 2](https://meetup.nhncloud.com/posts/92)

hyuntae99 replied:
근데 이러면  /api/review : write, /api/review/{review_id} : get, update, delete 이런 식으로 만들어야 하는데
SecurityConfig에 /api/review/*로 엔드포인트를 설정하면 get (리뷰 조회)를 할 때, 토큰을 요구하게 되더라고??
HTTP Method마다 제한을 걸어줄 수 있나?? (일단 get만 엔드포인트에 구분을 주었음)

seokho-1116 commented on src/main/java/com/project/chamjimayo/controller/ReviewController.java:
```diff
@@ -0,0 +1,260 @@
+	public ResponseEntity<ApiStandardResponse<ReviewDto>> updateReview(
+		@Parameter(description = "리뷰 ID", required = true, example = "1 (Long)")
+		@PathVariable Long reviewId,
+		@RequestBody ReviewDto reviewDto,
```
jwt 토큰으로 사용자 아이디를 받는데 ReviewDto에 왜 userId 필드가 포함되어 있는지?

hyuntae99 replied:
jwt 토큰으로 받아온 사용자 아이디랑 ReviewDto의 사용자 아이디가 같을 때만 
수정, 삭제가 가능하도록 만든건데 이렇게 구현하는게 아닌가?
(작성은 검증이 필요 없으니까 사용자 아이디가 없는 ReviewRequestDto로 받도록 하고)

hyuntae99 replied:
생각해보니까 필요 없을듯

seokho-1116 commented on src/main/java/com/project/chamjimayo/service/ReviewService.java:
```diff
@@ -0,0 +1,177 @@
+			.orElseThrow(
+				() -> new RestroomNotFoundException("해당 화장실을 찾을 수 없습니다. ID: " + restroomId));
+
+		Review review = new Review(user, restroom, reviewContent, rating);
```
new 말고 정적 팩토리 메서드는 어떤지? 객체 생성 관리 및 코드 이해에 용이함

hyuntae99 replied:
오케 적용했어

seokho-1116 commented on src/main/java/com/project/chamjimayo/service/UserService.java:
```diff
@@ -16,32 +18,75 @@
+    int newPoint = requestDTO.getPoint();
+    user.addPoint(currentPoint, newPoint);
+
+    userJpaRepository.save(user);
```
Jpa 변경 감지에 대한 내용을 보면 좋을 듯?
엔티티를 조회해 오면 스냅샷을 찍어놔서 트랜잭션이 끝날 때 변경 내역을 확인해서 업데이트 해줌.
아무 코드를 작성하지 않고도 특정 필드 or 전체 필드를 변경 후 업데이트 가능
[jpa dirty checking](https://jojoldu.tistory.com/415)

hyuntae99 replied:
적용했으

seokho-1116 commented on src/main/java/com/project/chamjimayo/service/UserService.java:
```diff
@@ -16,32 +18,75 @@
+    }
+    user.deductPoint(currentPoint, deductionPoint);
+
+    userJpaRepository.save(user);
```
여기도 마찬가지 jpa dirty checking

hyuntae99 replied:
적용했으

seokho-1116 reviewed (changes requested):
굿굿 수고 많았으 커멘트 남겼어

hyuntae99 committed:
1. Review 정적 메소드 도입
2. Review 엔드포인트 변경

hyuntae99 committed:
jpa dirty checking 적용

hyuntae99 committed:
SecurityConfig 엔드포인트 수정

hyuntae99 committed:
코드 컨밴션 적용

hyuntae99 committed:
ReviewDto userId, restroomId 제거 및 관련 코드 수정

hyuntae99 committed:
search entity 변경
(click 삭제, latitude/longitude 추가)

hyuntae99 committed:
1. 검색 결과 반환에 위도, 경도 추가
2. 검색 기록 클릭 로직 변경 (클릭한 주소만 저장)
3. 검색 기록 삭제 기능 추가
4. 최근 검색 기록 (리스트로 반환으로 변경)

hyuntae99 committed:
upstream merge

hyuntae99 committed:
swagger 필드 설명 수정

hyuntae99 committed:
swagger 설명 수정

hyuntae99 committed:
null point exception 에러 해결 (코드 순서 변경)

hyuntae99 commented:
### 작업 사항
1. 리뷰 수정에 불필요한 필드 삭제 -> ReviewDto 수정 (userId, restroomId 삭제)  
2.  search entity 변경 (click 삭제 + latitude, longitude 추가) -> erd cloud에 반영 완료
==> 기존엔 모든 검색 기록을 저장 후 click 변수로 탐색했지만 DB 과부화 문제로 수정
3. 검색 기록 삭제 기능, 최근 검색 기록 리스트 반환 추가

swagger로 테스트 완료했고 upstream 충돌도 해결했어!

kick-sim reviewed (approved):
수고했어!! 주석이 자세해서 보기 좋네요
